// grommet-controls/es6/components/Colors/doc.js
import { describe, PropTypes } from 'react-desc';

export default (function (Element) {
  var DocumentedElement = describe(Element).description('A color selection element, with custom color palettes.\n      ').usage('\n      $ npm install grommet-controls\n \n      import { Colors } from \'grommet-controls\';\n\n      <Colors colors={...} />\n    ');

  DocumentedElement.propTypes = {
    colors: PropTypes.object.isRequired.description('Color palette for the user to choose a color from.'),
    color: PropTypes.string.description('The default selected color.'),
    onSelect: PropTypes.func.description('Called with an object containing the selected color, rowName and colorName.'),
    size: PropTypes.oneOf(['small', 'medium', 'large']).description('What size to make it.'),
    columns: PropTypes.number.description('The maximum number of colors per row, if left emty will be calculated automatically.'),
    wrap: PropTypes.bool.description('If true, will wrap any rows of colors that exceed the columns limit, otherwise will truncate larger rows.')
  };

  return DocumentedElement;
});
