// grommet-controls/themes/black.js
'use strict';

exports.__esModule = true;

var _object = require('grommet/utils/object');

var _base = require('./base');

var _base2 = _interopRequireDefault(_base);

var _colorsFromArray = require('./colorsFromArray');

var _colorsFromArray2 = _interopRequireDefault(_colorsFromArray);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var brandColor = '#000000';
var accentColors = ['#FF4081', '#4eadeb', '#56cf98', '#765bba'];
var neutralColors = ['#a4c400', '#006766', '#f3741c', '#8F6C53', '#e82121'];
var backgroundColor = '#5e5e5e';

var colors = {
  'active': {
    light: '#3a82b1',
    dark: '#52b7f9'
  },
  'background': '#444444',
  'brand': brandColor,
  'focus': accentColors[0],
  'control': {
    'dark': '#f8f8f8',
    'light': '#444444'
  }
};
(0, _colorsFromArray2.default)(colors, accentColors, 'accent');
(0, _colorsFromArray2.default)(colors, neutralColors, 'neutral');

exports.default = (0, _object.deepFreeze)((0, _object.deepMerge)(_base2.default, {
  global: {
    colors: colors,
    drop: {
      background: '#444444'
    },
    font: {
      family: "'Inconsolata', monospace",
      size: '16px'
    },
    hover: {
      background: {
        light: '#dddddd',
        dark: '#222222'
      },
      color: {
        light: '#333333',
        dark: '#ffffff'
      }
    },
    control: {
      border: {
        radius: '0px'
      }
    }
  },
  icon: {
    colors: colors
  },
  anchor: {
    color: {
      light: '#0a466e',
      dark: '#4eadeb'
    }
  },
  button: {
    border: {
      radius: '0px'
    },
    primary: {
      color: {
        light: '#0a466e',
        dark: '#4eadeb'
      }
    }
  },
  checkBox: {
    check: {
      radius: '0px'
    },
    toggle: {
      radius: '0px'
    }
  },
  layer: {
    background: backgroundColor
  },
  rangeInput: {
    track: {
      color: {
        'dark': '#3a82b1',
        'light': '#52b7f9'
      }
    }
  }
}));
