// grommet-controls/themes/materialdark.js
'use strict';

exports.__esModule = true;

var _object = require('grommet/utils/object');

var _base = require('./base');

var _base2 = _interopRequireDefault(_base);

var _colorsFromArray = require('./colorsFromArray');

var _colorsFromArray2 = _interopRequireDefault(_colorsFromArray);

var _colorsFromObject = require('./colorsFromObject');

var _colorsFromObject2 = _interopRequireDefault(_colorsFromObject);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var brandColor = '#000000';
var backgroundColor = '#444444';
var accentColors = ['#2AD2C9', '#FFC107', '#9C27B0', '#673AB7'];
var neutralColors = ['#795548', '#009688', '#8BC34A', '#CDDC39', '#FF9800'];
var statusColors = {
  critical: '#FF4081',
  error: '#F44336',
  warning: '#FFEB3B',
  ok: '#4CAF50',
  unknown: '#9E9E9E',
  disabled: '#9E9E9E'
};

var focusColor = accentColors[0];
var colors = {
  active: {
    light: '#f50057',
    dark: '#ff4081'
  },
  brand: brandColor,
  background: backgroundColor,
  focus: focusColor,
  control: {
    'dark': '#f8f8f8',
    'light': '#444444'
  }

};

(0, _colorsFromArray2.default)(colors, accentColors, 'accent');
(0, _colorsFromArray2.default)(colors, neutralColors, 'neutral');
(0, _colorsFromObject2.default)(colors, statusColors, 'status');

exports.default = (0, _object.deepFreeze)((0, _object.deepMerge)(_base2.default, {
  global: {
    colors: colors,
    drop: {
      background: '#2a2a2a'
    },
    hover: {
      background: {
        light: '#dddddd',
        dark: '#222222'
      },
      text: {
        light: '#000000',
        dark: '#f8f8f8'
      }
    },
    focus: {
      border: {
        color: '#f50057'
      }
    },
    font: {
      family: '"Roboto", "Helvetica", "Arial", sans-serif'
    },
    control: {
      border: {
        radius: '0px'
      }
    }
  },
  button: {
    border: {
      radius: '0px'
    },
    primary: {
      color: {
        light: '#f50057',
        dark: '#ff4081'
      }
    }
  },
  checkBox: {
    check: {
      radius: '0px'
    }
  },
  layer: {
    background: backgroundColor
  },
  anchor: {
    color: {
      light: '#f50057',
      dark: '#ff4081'
    }
  }
}));
