// grommet-controls/themes/metro.js
'use strict';

exports.__esModule = true;

var _object = require('grommet/utils/object');

var _base = require('./base');

var _base2 = _interopRequireDefault(_base);

var _colorsFromArray = require('./colorsFromArray');

var _colorsFromArray2 = _interopRequireDefault(_colorsFromArray);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var brandColor = '#0072c6';
var accentColors = ['#fa6800', '#128023', '#0050ef', '#d80073'];
var neutralColors = ['#a4c400', '#00aba9', '#BF5A15', '#8F6C53', '#9A1616'];

var colors = {
  'background': '#ffffff',
  'brand': brandColor,
  'control': {
    'dark': '#f8f8f8',
    'light': '#444444'
  }
};

(0, _colorsFromArray2.default)(colors, accentColors, 'accent');
(0, _colorsFromArray2.default)(colors, neutralColors, 'neutral');

exports.default = (0, _object.deepFreeze)((0, _object.deepMerge)(_base2.default, {
  global: {
    colors: colors,
    drop: {
      background: '#005696',
      border: {
        width: '0px',
        radius: '0px'
      }
    },
    hover: {
      background: {
        light: '#a5cdff',
        dark: '#405063'
      },
      color: {
        light: '#333333',
        dark: '#cccccc'
      }
    },
    focus: {
      border: {
        color: {
          light: brandColor,
          dark: '#003967'
        }
      }
    },
    font: {
      family: '-apple-system, system-ui, BlinkMacSystemFont, "Segoe UI", "Roboto", "Ubuntu", "Helvetica Neue", sans-serif'
    },
    control: {
      border: {
        radius: '0px'
      }
    }
  },
  button: {
    border: {
      radius: '0px'
    },
    primary: {
      color: {
        light: brandColor,
        dark: 'accent-1'
      }
    }
  },
  anchor: {
    color: {
      light: '#0078D4',
      dark: '#ffffff'
    }
  },
  checkBox: {
    border: {
      color: {
        light: 'rgba(0, 98, 186, 0.5)',
        dark: 'rgba(255, 255, 255, 0.5)'
      }
    },
    check: {
      radius: '0px'
    },
    color: {
      light: brandColor,
      dark: '#a6cfff'
    },
    toggle: {
      color: {
        dark: '#bdbdbd',
        light: brandColor
      },
      radius: '0px'
    }
  },
  layer: {
    backgroundColor: '#eef3fc',
    border: {
      radius: '0px'
    }
  },
  radioButton: {
    border: {
      color: {
        light: 'rgba(0, 98, 186, 0.5)',
        dark: 'rgba(255, 255, 255, 0.5)'
      }
    }
  }
}));
