// grommet-controls/components/Colors/doc.js
'use strict';

exports.__esModule = true;

var _reactDesc = require('react-desc');

exports.default = function (Element) {
  var DocumentedElement = (0, _reactDesc.describe)(Element).description('A color selection element, with custom color palettes.\n      ').usage('\n      $ npm install grommet-controls\n \n      import { Colors } from \'grommet-controls\';\n\n      <Colors colors={...} />\n    ');

  DocumentedElement.propTypes = {
    colors: _reactDesc.PropTypes.object.isRequired.description('Color palette for the user to choose a color from.'),
    color: _reactDesc.PropTypes.string.description('The default selected color.'),
    onSelect: _reactDesc.PropTypes.func.description('Called with an object containing the selected color, rowName and colorName.'),
    size: _reactDesc.PropTypes.oneOf(['small', 'medium', 'large']).description('What size to make it.'),
    columns: _reactDesc.PropTypes.number.description('The maximum number of colors per row, if left emty will be calculated automatically.'),
    wrap: _reactDesc.PropTypes.bool.description('If true, will wrap any rows of colors that exceed the columns limit, otherwise will truncate larger rows.')
  };

  return DocumentedElement;
};
