// grommet-controls/es6/components/Header/Header.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

import React from 'react';
import { StyledHeader } from './StyledHeader';

var Header = function Header(_ref) {
  var children = _ref.children,
      rest = _objectWithoutProperties(_ref, ['children']);

  return React.createElement(
    StyledHeader,
    _extends({
      tag: 'header'
    }, rest),
    children
  );
};

Header.defaultProps = {
  position: 'relative',
  zIndex: 10,
  direction: 'row',
  background: 'brand',
  fill: 'horizontal',
  justify: 'between',
  align: 'center',
  elevation: 'medium',
  pad: 'medium'
};

var HeaderDoc = void 0;
if (process.env.NODE_ENV !== 'production') {
  HeaderDoc = require('./doc').default(Header); // eslint-disable-line global-require
}

var HeaderWrapper = HeaderDoc || Header;

export { HeaderWrapper as Header };
