// grommet-controls/es6/components/uiColors/uiColors.js
export var aqua = '#00ffff';
export var azure = '#f0ffff';
export var beige = '#f5f5dc';
export var black = '#000000';
export var blue = '#0000ff';
export var brown = '#a52a2a';
export var cyan = '#00ffff';
export var darkblue = '#00008b';
export var darkcyan = '#008b8b';
export var darkgrey = '#a9a9a9';
export var darkgreen = '#006400';
export var darkkhaki = '#bdb76b';
export var darkmagenta = '#8b008b';
export var darkolivegreen = '#556b2f';
export var darkorange = '#ff8c00';
export var darkorchid = '#9932cc';
export var darkred = '#8b0000';
export var darksalmon = '#e9967a';
export var darkviolet = '#9400d3';
export var fuchsia = '#ff00ff';
export var gold = '#ffd700';
export var green = '#008000';
export var indigo = '#4b0082';
export var khaki = '#f0e68c';
export var lightblue = '#add8e6';
export var lightcyan = '#e0ffff';
export var lightgreen = '#90ee90';
export var lightgrey = '#d3d3d3';
export var lightpink = '#ffb6c1';
export var lightyellow = '#ffffe0';
export var lime = '#00ff00';
export var magenta = '#ff00ff';
export var maroon = '#800000';
export var navy = '#000080';
export var olive = '#808000';
export var orange = '#ffa500';
export var pink = '#ffc0cb';
export var purple = '#800080';
export var violet = '#800080';
export var red = '#ff0000';
export var silver = '#c0c0c0';
export var white = '#ffffff';
export var yellow = '#ffff00';

export var uiColors = {
  aqua: aqua,
  azure: azure,
  beige: beige,
  black: black,
  blue: blue,
  brown: brown,
  cyan: cyan,
  darkblue: darkblue,
  darkcyan: darkcyan,
  darkgrey: darkgrey,
  darkgreen: darkgreen,
  darkkhaki: darkkhaki,
  darkmagenta: darkmagenta,
  darkolivegreen: darkolivegreen,
  darkorange: darkorange,
  darkorchid: darkorchid,
  darkred: darkred,
  darksalmon: darksalmon,
  darkviolet: darkviolet,
  fuchsia: fuchsia,
  gold: gold,
  green: green,
  indigo: indigo,
  khaki: khaki,
  lightblue: lightblue,
  lightcyan: lightcyan,
  lightgreen: lightgreen,
  lightgrey: lightgrey,
  lightpink: lightpink,
  lightyellow: lightyellow,
  lime: lime,
  magenta: magenta,
  maroon: maroon,
  navy: navy,
  olive: olive,
  orange: orange,
  pink: pink,
  purple: purple,
  violet: violet,
  red: red,
  silver: silver,
  white: white,
  yellow: yellow
};
