// grommet-controls/components/PagingTable/StyledPagingTable.js
'use strict';

exports.__esModule = true;
exports.StyledPagingTable = exports.StyledTfootComponent = exports.StyledExpander = exports.StyledTdComponent = exports.StyledTrComponent = exports.StyledTrGroupComponent = exports.StyledThComponent = exports.StyledResizerComponent = exports.StyledTHeadComponent = exports.StyledTableBodyComponent = exports.StyledTableComponent = undefined;

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _grommet = require('grommet');

var _colors = require('grommet/utils/colors');

var _ReactTable = require('./react-table/ReactTable');

var _ReactTable2 = _interopRequireDefault(_ReactTable);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var StyledTableComponent = exports.StyledTableComponent = (0, _styledComponents2.default)(_grommet.Box).withConfig({
  displayName: 'StyledPagingTable__StyledTableComponent',
  componentId: 'sc-1wbkk7c-0'
})(['align-items:stretch;width:100%;border-collapse:collapse;overflow:auto']);

var StyledTableBodyComponent = exports.StyledTableBodyComponent = (0, _styledComponents2.default)(_grommet.Box).withConfig({
  displayName: 'StyledPagingTable__StyledTableBodyComponent',
  componentId: 'sc-1wbkk7c-1'
})(['']);

var StyledTHeadComponent = exports.StyledTHeadComponent = (0, _styledComponents2.default)(_grommet.Box).withConfig({
  displayName: 'StyledPagingTable__StyledTHeadComponent',
  componentId: 'sc-1wbkk7c-2'
})(['-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;']);

var StyledResizerComponent = exports.StyledResizerComponent = (0, _styledComponents2.default)(_grommet.Box).withConfig({
  displayName: 'StyledPagingTable__StyledResizerComponent',
  componentId: 'sc-1wbkk7c-3'
})(['display:inline-block;position:absolute;width:36px;top:0;bottom:0;right:-18px;cursor:col-resize;z-index:10;']);

var StyledThComponent = exports.StyledThComponent = (0, _styledComponents2.default)(_grommet.Box).withConfig({
  displayName: 'StyledPagingTable__StyledThComponent',
  componentId: 'sc-1wbkk7c-4'
})(['position:relative;', '     ', ' ', '   ', '  '], function (props) {
  return props.sortable && '\n    cursor: pointer;\n  ';
}, function (props) {
  return props.hidden && '\n    width: 0 !important;\n    min-width: 0 !important;\n    padding: 0 !important;\n    border: 0 !important;\n    opacity: 0 !important;\n  ';
}, function (props) {
  return props.pivot && '\n    &:after,:before {\n      left: 100%;\n      top: 50%;\n      border: solid transparent;\n      content: " ";\n      height: 0;\n      width: 0;\n      position: absolute;\n      pointer-events: none\n    }\n  \n    &:after {\n      border-color: rgba(255, 255, 255, 0);\n      border-left-color: #fff;\n      border-width: 8px;\n      margin-top: -8px\n    }\n  \n    &:before {\n      border-color: rgba(102, 102, 102, 0);\n      border-left-color: #f7f7f7;\n      border-width: 10px;\n      margin-top: -10px\n    }\n\n  ';
}, function (props) {
  return props.resizable && '\n    overflow: visible;\n\n    &:last-child {\n      overflow: hidden\n    }\n  ';
});

var StyledTrGroupComponent = exports.StyledTrGroupComponent = (0, _styledComponents2.default)(_grommet.Box).withConfig({
  displayName: 'StyledPagingTable__StyledTrGroupComponent',
  componentId: 'sc-1wbkk7c-5'
})(['align-items:stretch;']);

var StyledTrComponent = exports.StyledTrComponent = (0, _styledComponents2.default)(_grommet.Box).withConfig({
  displayName: 'StyledPagingTable__StyledTrComponent',
  componentId: 'sc-1wbkk7c-6'
})(['', ''], function (props) {
  return props.hover && '\n    &:hover {\n      background-color: ' + (0, _colors.normalizeColor)(props.hover.background || props.theme.global.hover.background, props.theme) + ';\n      color: ' + (0, _colors.normalizeColor)(props.hover.color || props.theme.global.hover.color, props.theme) + ';\n    }  \n  ';
});

var StyledTdComponent = exports.StyledTdComponent = (0, _styledComponents2.default)(_grommet.Box).withConfig({
  displayName: 'StyledPagingTable__StyledTdComponent',
  componentId: 'sc-1wbkk7c-7'
})(['text-overflow:ellipsis;overflow:hidden;display:block;white-space:nowrap;', ''], function (props) {
  return props.hidden && '\n    width: 0 !important;\n    min-width: 0 !important;\n    padding: 0 !important;\n    border: 0 !important;\n    opacity: 0 !important\n  ';
});

var StyledExpander = exports.StyledExpander = (0, _styledComponents2.default)(_grommet.Button).withConfig({
  displayName: 'StyledPagingTable__StyledExpander',
  componentId: 'sc-1wbkk7c-8'
})(['position:relative;margin:0;padding:0;left:-5px;color:transparent;cursor:pointer;']);

var StyledTfootComponent = exports.StyledTfootComponent = (0, _styledComponents2.default)(_grommet.Box).withConfig({
  displayName: 'StyledPagingTable__StyledTfootComponent',
  componentId: 'sc-1wbkk7c-9'
})(['']);

var StyledPagingTable = exports.StyledPagingTable = (0, _styledComponents2.default)(_ReactTable2.default).withConfig({
  displayName: 'StyledPagingTable',
  componentId: 'sc-1wbkk7c-10'
})(['max-width:100%;width:100%;overflow:hidden;']);
