// grommet-controls/components/PagingTable/components/ThComponent.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _grommet = require('grommet');

var _Blank = require('grommet-icons/icons/Blank');

var _LinkDown = require('grommet-icons/icons/LinkDown');

var _LinkUp = require('grommet-icons/icons/LinkUp');

var _StyledPagingTable = require('../StyledPagingTable');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

exports.default = function (_ref) {
  var toggleSort = _ref.toggleSort,
      sort = _ref.sort,
      resizable = _ref.resizable,
      children = _ref.children,
      pivot = _ref.pivot,
      hidden = _ref.hidden,
      sortable = _ref.sortable,
      expander = _ref.expander,
      CellTextComponent = _ref.CellTextComponent,
      props = _objectWithoutProperties(_ref, ['toggleSort', 'sort', 'resizable', 'children', 'pivot', 'hidden', 'sortable', 'expander', 'CellTextComponent']);

  if (!expander && Array.isArray(children) && children.length > 1 && !children[0] && !children[1]) {
    return null;
  }

  var style = props.style,
      rest = _objectWithoutProperties(props, ['style']);

  var childRendered = _react2.default.createElement(CellTextComponent, _extends({}, rest, { value: children }));
  var content = void 0;
  if (sortable) {
    var sortAsc = sort === '-sort-asc';
    var sortDesc = sort === '-sort-desc';
    var SortIcon = _Blank.Blank;
    if (sortAsc || sortDesc) {
      SortIcon = sortAsc ? _LinkUp.LinkUp : _LinkDown.LinkDown;
    }
    content = _react2.default.createElement(
      _grommet.Box,
      { direction: 'row', gap: 'xsmall', align: 'center' },
      childRendered,
      _react2.default.createElement(
        _grommet.Button,
        { onClick: function onClick(e) {
            return toggleSort && toggleSort(e);
          } },
        _react2.default.createElement(SortIcon, null)
      )
    );
  } else {
    content = childRendered;
  }

  var boxProps = _extends({ pad: 'xsmall' }, rest);
  return (

    // eslint-disable-next-line jsx-a11y/interactive-supports-focus
    _react2.default.createElement(
      _StyledPagingTable.StyledThComponent,
      {
        role: 'columnheader',
        onClick: function onClick(e) {
          return toggleSort && toggleSort(e);
        },
        direction: 'row',
        pivot: pivot,
        hidden: hidden,
        sortable: sortable,
        style: style,
        resizable: resizable
      },
      _react2.default.createElement(
        _grommet.Box,
        _extends({}, boxProps, { fill: true }),
        content
      )
    )
  );
};
