// grommet-controls/components/uiColors/uiColors.js
'use strict';

exports.__esModule = true;
var aqua = exports.aqua = '#00ffff';
var azure = exports.azure = '#f0ffff';
var beige = exports.beige = '#f5f5dc';
var black = exports.black = '#000000';
var blue = exports.blue = '#0000ff';
var brown = exports.brown = '#a52a2a';
var cyan = exports.cyan = '#00ffff';
var darkblue = exports.darkblue = '#00008b';
var darkcyan = exports.darkcyan = '#008b8b';
var darkgrey = exports.darkgrey = '#a9a9a9';
var darkgreen = exports.darkgreen = '#006400';
var darkkhaki = exports.darkkhaki = '#bdb76b';
var darkmagenta = exports.darkmagenta = '#8b008b';
var darkolivegreen = exports.darkolivegreen = '#556b2f';
var darkorange = exports.darkorange = '#ff8c00';
var darkorchid = exports.darkorchid = '#9932cc';
var darkred = exports.darkred = '#8b0000';
var darksalmon = exports.darksalmon = '#e9967a';
var darkviolet = exports.darkviolet = '#9400d3';
var fuchsia = exports.fuchsia = '#ff00ff';
var gold = exports.gold = '#ffd700';
var green = exports.green = '#008000';
var indigo = exports.indigo = '#4b0082';
var khaki = exports.khaki = '#f0e68c';
var lightblue = exports.lightblue = '#add8e6';
var lightcyan = exports.lightcyan = '#e0ffff';
var lightgreen = exports.lightgreen = '#90ee90';
var lightgrey = exports.lightgrey = '#d3d3d3';
var lightpink = exports.lightpink = '#ffb6c1';
var lightyellow = exports.lightyellow = '#ffffe0';
var lime = exports.lime = '#00ff00';
var magenta = exports.magenta = '#ff00ff';
var maroon = exports.maroon = '#800000';
var navy = exports.navy = '#000080';
var olive = exports.olive = '#808000';
var orange = exports.orange = '#ffa500';
var pink = exports.pink = '#ffc0cb';
var purple = exports.purple = '#800080';
var violet = exports.violet = '#800080';
var red = exports.red = '#ff0000';
var silver = exports.silver = '#c0c0c0';
var white = exports.white = '#ffffff';
var yellow = exports.yellow = '#ffff00';

var uiColors = exports.uiColors = {
  aqua: aqua,
  azure: azure,
  beige: beige,
  black: black,
  blue: blue,
  brown: brown,
  cyan: cyan,
  darkblue: darkblue,
  darkcyan: darkcyan,
  darkgrey: darkgrey,
  darkgreen: darkgreen,
  darkkhaki: darkkhaki,
  darkmagenta: darkmagenta,
  darkolivegreen: darkolivegreen,
  darkorange: darkorange,
  darkorchid: darkorchid,
  darkred: darkred,
  darksalmon: darksalmon,
  darkviolet: darkviolet,
  fuchsia: fuchsia,
  gold: gold,
  green: green,
  indigo: indigo,
  khaki: khaki,
  lightblue: lightblue,
  lightcyan: lightcyan,
  lightgreen: lightgreen,
  lightgrey: lightgrey,
  lightpink: lightpink,
  lightyellow: lightyellow,
  lime: lime,
  magenta: magenta,
  maroon: maroon,
  navy: navy,
  olive: olive,
  orange: orange,
  pink: pink,
  purple: purple,
  violet: violet,
  red: red,
  silver: silver,
  white: white,
  yellow: yellow
};
