// grommet-controls/components/Form/doc.js
'use strict';

exports.__esModule = true;

var _reactDesc = require('react-desc');

var _propTypes = require('grommet/utils/prop-types');

exports.default = function (Element) {
  var DocumentedElement = (0, _reactDesc.describe)(Element).description('A Form with built-in validation.\n      ').usage('\n      $ npm install grommet-controls\n \n      import { Form } from \'grommet-controls\';\n\n      <Form onSubmit={...} />\n    ');

  DocumentedElement.propTypes = {
    a11yTitle: _propTypes.a11yTitlePropType,
    basis: _reactDesc.PropTypes.oneOf(['xsmall', 'small', 'medium', 'large', 'xlarge', 'full', '1/2', '1/3', '2/3', '1/4', '3/4', 'auto']).description('A fixed or relative size along the form\'s container main axis.'),
    focusFirstChild: _reactDesc.PropTypes.bool.description('Whether to focus the first form control on mounting.').defaultValue('true'),
    object: _reactDesc.PropTypes.object.description('Initial values of the form data object.'),
    onSubmit: _reactDesc.PropTypes.func.description('A function called when the user successfully submits the form.'),
    onChange: _reactDesc.PropTypes.func.description('A function invoked when any of the values in the form are changing.'),
    onSubmitError: _reactDesc.PropTypes.func.description('A function invoked when an error occured during submitting.'),
    onInvalidForm: _reactDesc.PropTypes.func.description('A function invoked on every change when the form is invalid.'),
    tag: _reactDesc.PropTypes.string.description('The DOM tag to use for the form.').defaultValue('form')
  };

  return DocumentedElement;
};
