// grommet-controls/components/Value/Value.js
'use strict';

exports.__esModule = true;
exports.Value = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _grommet = require('grommet');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var LABEL_SIZE_MAP = {
  xsmall: 'xsmall',
  small: 'xsmall',
  medium: 'xsmall',
  large: 'small',
  xlarge: 'medium',
  xxlarge: 'large'
};
var Value = function Value(_ref) {
  var label = _ref.label,
      value = _ref.value,
      size = _ref.size,
      weight = _ref.weight,
      gap = _ref.gap,
      units = _ref.units,
      color = _ref.color;
  return _react2.default.createElement(
    _grommet.Box,
    { gap: gap },
    _react2.default.isValidElement(value) ? value : _react2.default.createElement(
      _grommet.Text,
      { size: size, weight: weight, color: color },
      '' + (value !== undefined ? value : '-') + (units ? ' ' + units : '')
    ),
    _react2.default.createElement(
      _grommet.Box,
      { alignSelf: 'end' },
      _react2.default.createElement(
        _grommet.Text,
        { size: LABEL_SIZE_MAP[size], color: color },
        label
      )
    )
  );
};

Value.defaultProps = {
  label: undefined,
  value: undefined,
  weight: 'bold',
  gap: 'small',
  size: 'large',
  color: undefined
};

Value.propTypes = {
  label: _propTypes2.default.string,
  color: _propTypes2.default.string,
  value: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.element, _propTypes2.default.number]),
  weight: _propTypes2.default.oneOfType([_propTypes2.default.oneOf(['normal', 'bold']), _propTypes2.default.number]),
  gap: _propTypes2.default.oneOf(['xsmall', 'small', 'medium', 'large', 'xlarge']),
  size: _propTypes2.default.oneOf(['xsmall', 'small', 'medium', 'large', 'xlarge', 'xxlarge'])
};

var ValueDoc = void 0;
if (process.env.NODE_ENV !== 'production') {
  ValueDoc = require('./doc').default(Value); // eslint-disable-line global-require
}

var ValueWrapper = ValueDoc || Value;

exports.Value = ValueWrapper;
