// grommet-controls/components/Header/Header.js
'use strict';

exports.__esModule = true;
exports.Header = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _StyledHeader = require('./StyledHeader');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var Header = function Header(_ref) {
  var children = _ref.children,
      rest = _objectWithoutProperties(_ref, ['children']);

  return _react2.default.createElement(
    _StyledHeader.StyledHeader,
    _extends({
      tag: 'header'
    }, rest),
    children
  );
};

Header.defaultProps = {
  position: 'relative',
  zIndex: 10,
  direction: 'row',
  background: 'brand',
  fill: 'horizontal',
  justify: 'between',
  align: 'center',
  elevation: 'medium',
  pad: 'medium'
};

var HeaderDoc = void 0;
if (process.env.NODE_ENV !== 'production') {
  HeaderDoc = require('./doc').default(Header); // eslint-disable-line global-require
}

var HeaderWrapper = HeaderDoc || Header;

exports.Header = HeaderWrapper;
