// grommet-controls/components/MaskedInput/doc.js
'use strict';

exports.__esModule = true;

var _reactDesc = require('react-desc');

var _propTypes = require('grommet/utils/prop-types');

exports.default = function (Element) {
  var DocumentedElement = (0, _reactDesc.describe)(Element).description('A masked Input control with an optional drop button with the specified \'dropContent\'.').usage('\n    $ npm install grommet-controls\n\n    import { MakedInput } from \'grommet-controls\';\n\n    <MakedInput mask={...} />\n    ');

  DocumentedElement.propTypes = {
    a11yTitle: _propTypes.a11yTitlePropType,
    a11yDropTitle: _reactDesc.PropTypes.string.description('Custom drop button title to be used by screen readers.'),
    placeholder: _reactDesc.PropTypes.string.description('Placeholder text to use when no value is provided.'),
    value: _reactDesc.PropTypes.oneOfType([_reactDesc.PropTypes.string, _reactDesc.PropTypes.number]).description('Value for the field.'),
    id: _reactDesc.PropTypes.string.description('The id attribute of the input.'),
    name: _reactDesc.PropTypes.string.description('The name attribute of the input.'),
    disabled: _reactDesc.PropTypes.bool.description('Setting to true causes the input to be disabled.'),
    onChange: _reactDesc.PropTypes.func.description('Function that will be called when the user enters a new valuu.'),
    focusIndicator: _reactDesc.PropTypes.bool.description('Whether the plain text input should receive a focus outline.'),
    plain: _reactDesc.PropTypes.bool.description('Whether this is a plain input with no border or padding.\nOnly use this when the containing context provides sufficient affordance'),
    mask: _reactDesc.PropTypes.oneOfType([_reactDesc.PropTypes.array, _reactDesc.PropTypes.func, _reactDesc.PropTypes.bool, _reactDesc.PropTypes.shape({
      mask: _reactDesc.PropTypes.oneOfType([_reactDesc.PropTypes.array, _reactDesc.PropTypes.func]),
      pipe: _reactDesc.PropTypes.func
    })]).description('An array or a function that defines how the user input is going to be masked.'),
    guide: _reactDesc.PropTypes.bool.description('When masked, a boolean that tells the component whether to be in guide or no guide mode.'),
    pipe: _reactDesc.PropTypes.func.description('When masked, a function that will give you the opportunity to modify the conformed value before it is displayed on the screen.'),
    placeholderChar: _reactDesc.PropTypes.string.description('When masked, represents the fillable spot in the mask.'),
    keepCharPositions: _reactDesc.PropTypes.bool.description('For masked. When true, adding or deleting characters will not affect the positions of existing characters.\n      When false, adding characters causes existing characters to advance. And deleting characters causes existing characters to move back.'),
    showMask: _reactDesc.PropTypes.bool.description('When masked, displays the mask as a placeholder in place of the regular placeholder.'),
    dropContent: _reactDesc.PropTypes.element.description('Content to put inside the Drop.'),
    dropIcon: _reactDesc.PropTypes.element.description('Icon for drop content.'),
    widgets: _reactDesc.PropTypes.arrayOf(_reactDesc.PropTypes.shape({
      icon: _reactDesc.PropTypes.element.isRequired
    })).description('Additional widgets to be placed next to the input.'),
    onClose: _reactDesc.PropTypes.func.description('Callback for when the drop is closed')
  };

  return DocumentedElement;
};
