// grommet-controls/components/DateInput/doc.js
'use strict';

exports.__esModule = true;

var _reactDesc = require('react-desc');

exports.default = function (Element) {
  var DocumentedElement = (0, _reactDesc.describe)(Element).description('A masked date entry field with a drop down calendar.\n      All properties of MaskedInput apply\n      ').usage('\n    $ npm install grommet-controls\n\n    import { DateInput } from \'grommet-controls\';\n\n    <DateInput value={...} />\n    ');

  DocumentedElement.propTypes = {
    autocorrect: _reactDesc.PropTypes.bool.description('Use mask that wil attempt to auto-correct the entry as the user types'),
    disabled: _reactDesc.PropTypes.bool.description('Setting to true causes the input to be disabled.'),
    bounds: _reactDesc.PropTypes.arrayOf(_reactDesc.PropTypes.string).description('An array of two numbers indicating the limits on\n        navigation in ISO8601 format'),
    disabledDates: _reactDesc.PropTypes.arrayOf(_reactDesc.PropTypes.oneOfType([_reactDesc.PropTypes.string, _reactDesc.PropTypes.arrayOf(_reactDesc.PropTypes.string)])).description('Multiple dates in ISO8601 format that should not be\n        selectable. Items that are an array indicate a range of dates.'),
    firstDayOfWeek: _reactDesc.PropTypes.oneOf([0, 1]).description('The first day of the week. 0 for Sunday. 1 for Monday.'),
    locale: _reactDesc.PropTypes.string.description('The locale to use.'),
    size: _reactDesc.PropTypes.oneOf(['small', 'medium', 'large']).description('Size of the calendar drop.'),
    value: _reactDesc.PropTypes.oneOfType([_reactDesc.PropTypes.instanceOf(Date), _reactDesc.PropTypes.string]).description('The date value is Date format of ISO string')
  };

  return DocumentedElement;
};
